File: Python/49.py
# Notes
    # sorting is slow, so want to use a hash
    # can't use Counter() for hash keys because they're mutable
    # instead, create tuple version of counter by making list of char counts
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        d = defaultdict(list)
        
        for word in strs:
            count = [0]*26
            for char in word:
                count[ord(char) - ord("a")] += 1
            d[tuple(count)].append(word)

        return list(d.values())

File: C++/49.cpp
/*
- Easy way: use sorted version of string as key for hash
- Better way: use counter of string as hash key
*/
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> map;
        for(int i = 0; i < strs.size(); i++){
            string key = get_key(strs[i]);
            cout << key << endl;
            map[key].push_back(strs[i]);
        }
        vector<vector<string>> result;
        for(auto it = map.begin(); it != map.end(); it++){
            result.push_back(it -> second);
        }
        return result;
    }
private:
    string get_key(string str){
        vector<int> counter(26);
        for(int i = 0; i < str.size(); i++){
            counter[str[i] - 'a']++;
        }
        string key = "";
        for(int i = 0; i < counter.size(); i++){
            key.append(to_string(counter[i]) + "#");
        }
        return key;
    }
};